File: ocaml/procedural/pid.ml
type state = {
  mutable position : float;
  mutable velocity : float;
  mutable integral : float;
  mutable prev_error : float;
}

type params = {
  target : float;
  mass : float;
  k_p : float;
  k_i : float;
  k_d : float;
  dt : float;
  total_steps : int;
}

let simulate_pid state target mass k_p k_i k_d dt total_steps =
  for step = 0 to total_steps - 1 do
    let error = target -. state.position in
    let derivative = (error -. state.prev_error) /. dt in
    let force = (k_p *. error) +. (k_i *. state.integral) +. (k_d *. derivative) in
    let acceleration = force /. mass in

    state.velocity <- state.velocity +. acceleration *. dt;
    state.position <- state.position +. state.velocity *. dt;
    state.integral <- state.integral +. error *. dt;
    state.prev_error <- error
  done

let () =
  let initial_state = { position = 0.0; velocity = 0.0; integral = 0.0; prev_error = 0.0 } in
  let params = {
    target = 10.0;
    mass = 2.0;
    k_p = 0.5;
    k_i = 0.1;
    k_d = 0.2;
    dt = 0.1;
    total_steps = 1000
  } in
  
  simulate_pid initial_state params.target params.mass params.k_p params.k_i params.k_d params.dt params.total_steps;
  Printf.printf "Final position: %.2f\n" initial_state.position

File: rust/procedural/pid.rs
struct State {
    position: f64,
    velocity: f64,
    integral: f64,
    prev_error: f64,
}

fn simulate_pid(
    state: &mut State,
    target: f64,
    mass: f64,
    k_p: f64,
    k_i: f64,
    k_d: f64,
    dt: f64,
    total_steps: usize,
) {
    for step in 0..total_steps {
        let error = target - state.position;
        let derivative = (error - state.prev_error) / dt;
        let force = k_p * error + k_i * state.integral + k_d * derivative;
        
        let acceleration = force / mass;
        state.velocity += acceleration * dt;
        state.position += state.velocity * dt;
        state.integral += error * dt;
        state.prev_error = error;

        // Uncomment for debugging:
        // println!("Step {}: Position = {:.2}", step + 1, state.position);
    }
}

fn main() {
    let mut state = State {
        position: 0.0,
        velocity: 0.0,
        integral: 0.0,
        prev_error: 0.0,
    };
    
    simulate_pid(
        &mut state,
        10.0,   // Target
        2.0,    // Mass
        0.5, 0.1, 0.2, // PID gains
        0.1,    // dt
        1000,   // Steps
    );
    
    println!("Final position: {:.2}", state.position);
}